tests/scripts/all.sh

```diff
@@ -644,6 +647,28 @@ pre_check_tools () {
     "$@" scripts/output_env.sh
 }
 
+# Set up an include directory for baremetal tests. Exclude headers that the
+# configuration says we shouldn't rely upon.
+pre_setup_baremetal_usr_include () {
+    msg "info: setting up include directory for bare metal tests"
+    case " $RUN_COMPONENTS " in
+        *baremetal*)
+            cp -r /usr/include baremetal-usr-include
+            rm baremetal-usr-include/time.h
+            rm baremetal-usr-include/pthread.h
+            ;;
+    esac
+
+    # We also need a set of CFLAGS to use this. We need to preserve the
+    # compiler's other system include directories, preferably without
+    # assuming too much about the compiler or platform. However, we
+    # assume gcc for now.
+
+    dirs=$(gcc -xc -E -v - < /dev/null 2>&1 | egrep '^ /usr/.*include.*$' | \
+           egrep -v '/usr/include$' | sed 's/^/ -isystem/' | tr -d "\n")
+
+    BAREMETAL_INCLUDE_OVERRIDE="-nostdinc $dirs -I $(pwd)/baremetal-usr-include"
```

Comment:
Just my two cents on this:
The approach seems complicated and error-prone as it relies on a specific compiler and assumes it is run on a Unix-like platform.
Let me propose an alternative: Do the opposite: don't "remove" the header files, change the `include` statements to reference something that (most probable) does not exist? For example
```
#include <time.h>
```
would be changed to
```
#include <time_h_marked_not_existant_by_config>
```
One option to do this, is to use another layer of indirection: Change the literal file names into macro like:
```
#include <config.h>
#include <TIME_H>
```
where `config.h` contains:
```
#ifdef MBEDTLS_HAVE_TIME
#define TIME_H time.h
#else 
#define TIME_H time_h_marked_not_existant_by_config
#endif
```
Maybe someone of the maintainers (e.g., @gilles-peskine-arm) could weigh in here to avoid unnecessary work.

Reply:
It seems to me that it would be simpler to create a `tests/include/no-time/time.h` containing `#error "time.h was included but should not have been"`, and add that as a `-I` option for the component that tests without `MBEDTLS_HAVE_TIME`.

I don't like changing the `#include` statements: it adds complexity to the code for the sake of one test. Adding another layer of indirection is not acceptable because expanding macros in `#include` statements is non-standard; GCC and Clang do it but some other compilers choke on it. We do it for `MBEDTLS_CONFIG_FILE`, but as a consequence `MBEDTLS_CONFIG_FILE` only works with compilers that support that non-standard feature.

Reply:
Oh, `-I tests/include/something` is a fantastic suggestion. I will do that, that's _much_ better. Thanks.

Reply:
Agreed! A terrific find! :+1:

Reply:
New approach adds `tests/include/baremetal-override/time.h`. I've picked `baremetal-override` so we can experiment with things like disabling `MBEDTLS_FS_IO` in future - that one has bitten me a couple of times.